src/main/java/org/apache/xml/security/utils/DOMNamespaceContext.java

```diff
@@ -18,58 +18,124 @@
  */
 package org.apache.xml.security.utils;
 
-import java.util.HashMap;
-import java.util.Iterator;
-import java.util.Map;
-import java.util.Map.Entry;
+import org.w3c.dom.Node;
 
 import javax.xml.namespace.NamespaceContext;
+import java.util.Iterator;
+import java.util.Objects;
 
-import org.w3c.dom.Attr;
-import org.w3c.dom.Element;
-import org.w3c.dom.NamedNodeMap;
-import org.w3c.dom.Node;
+import static javax.xml.XMLConstants.DEFAULT_NS_PREFIX;
+import static javax.xml.XMLConstants.NULL_NS_URI;
+import static javax.xml.XMLConstants.XMLNS_ATTRIBUTE;
+import static javax.xml.XMLConstants.XMLNS_ATTRIBUTE_NS_URI;
+import static javax.xml.XMLConstants.XML_NS_PREFIX;
+import static javax.xml.XMLConstants.XML_NS_URI;
 
 /**
+ * This class adapts the {@link Node} namespace/prefix lookup API to that of {@link NamespaceContext}.
+ * There are some differences:
+ * <table>
+ *     <tr>
+ *         <th>Function</th>
+ *         <th>{@link NamespaceContext} API</th>
+ *         <th>{@link Node} API</th>
+ *     </tr>
+ *     <tr>
+ *         <td>Look up the prefix for a given namespace URI.</td>
+ *         <td>{@link NamespaceContext#getPrefix(String)}</td>
+ *         <td>{@link Node#lookupPrefix(String)}</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Look up all the prefixes for a given namespace URI.</td>
+ *         <td>{@link NamespaceContext#getPrefixes(String)}</td>
+ *         <td>/</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Look up the namespace URI for a given prefix.</td>
+ *         <td>{@link NamespaceContext#getNamespaceURI(String)}</td>
+ *         <td>{@link Node#lookupNamespaceURI(String)}</td>
+ *     </tr>
+ *     <tr>
+ *         <td>The default prefix.</td>
+ *         <td>{@link javax.xml.XMLConstants#DEFAULT_NS_PREFIX}</td>
+ *         <td>{@code null}</td>
+ *     </tr>
+ *     <tr>
+ *         <td>The default namespace URI.</td>
+ *         <td>{@link javax.xml.XMLConstants#NULL_NS_URI}</td>
+ *         <td>{@code null}</td>
+ *     </tr>
+ * </table>
  */
 public class DOMNamespaceContext implements NamespaceContext {
 
-    private Map<String, String> namespaceMap = new HashMap<>();
+    private Node context;
 
-    public DOMNamespaceContext(Node contextNode) {
-        addNamespaces(contextNode);
+    public DOMNamespaceContext(Node context) {
+        setContext(context);
     }
 
-    public String getNamespaceURI(String arg0) {
-        return namespaceMap.get(arg0);
+    public void setContext(Node context) {
+        this.context = context;
     }
 
-    public String getPrefix(String arg0) {
-        for (Entry<String, String> entry : namespaceMap.entrySet()) {
-            if (entry.getValue().equals(arg0)) {
-                return entry.getKey();
+    public String getNamespaceURI(String prefix) {
```

Comment:
For getNamespaceURI, I'm wondering if it should be falling back to returning "" instead of return null? From the javadoc:

unbound prefix | XMLConstants.NULL_NS_URI("")

Reply:
You have a point there...
Still, I'd like to explain why I implemented it this way on purpose: returning `null` for an unbound namespace allows the caller to distinguish between "unbound prefix" and "prefix for the default namespace", which could be useful.
Collapsing both "unbound prefix" and "prefix for the default namespace" preempts the caller from being able to make that distinction.
Of course, my opinion on "what the API should have been" doesn't really matter here, I have to obey the contract of `getNamespaceURI`.
I'll do a little bit more research (and change several of my unit tests), but I'll probably change it according to your suggestion.